### PR TITLE
feat: [NODE-1441] Upgrade GuestOS to 24.04

### PR DESCRIPTION
**Pull request by Bownairo**
Switch GuestOS over to a Ubuntu 24.04 base image. These changes will be initially rolled out in an RC.

### CODE REVIEW

**Review by marko-k0 — APPROVED** [review]

**Review by andrewbattat — APPROVED** [review]

File: ic-os/components/misc/guestos/crypttab [hunk 1]
```diff
@@ -1 +1 @@
-vda10-crypt /dev/vda10  /boot/config/store.keyfile  luks2,discard
+vda10-crypt /dev/vda10  /boot/config/store.keyfile  luks,discard
```
[the remark this answers is not in the document]
Comment (Bownairo) [reply ic-os/components/misc/guestos/crypttab:1]: I don't think `luks2` has ever been a valid setting for [cryptab](https://www.man7.org/linux/man-pages/man5/crypttab.5.html), but it has been smart enough to detect luks on its own. This just silences a warning.

File: ic-os/defs.bzl [hunk 2]
```diff
@@ -147,8 +147,15 @@ def icos_build(
...
-            "/usr/lib/firmware/brcm/brcmfmac43455-sdio.MINIX-NEO Z83-4.txt",
+            "/usr/lib/firmware/brcm/brcmfmac43241b4-sdio.Intel Corp.-VALLEYVIEW C0 PLATFORM.txt.zst",
+            "/usr/lib/firmware/brcm/brcmfmac43340-sdio.ASUSTeK COMPUTER INC.-TF103CE.txt.zst",
+            "/usr/lib/firmware/brcm/brcmfmac43362-sdio.ASUSTeK COMPUTER INC.-ME176C.txt.zst",
+            "/usr/lib/firmware/brcm/brcmfmac43430a0-sdio.ONDA-V80 PLUS.txt.zst",
+            "/usr/lib/firmware/brcm/brcmfmac43455-sdio.MINIX-NEO Z83-4.txt.zst",
+            "/usr/lib/firmware/brcm/brcmfmac43455-sdio.Raspberry Pi Foundation-Raspberry Pi 4 Model B.txt.zst",
+            "/usr/lib/firmware/brcm/brcmfmac43455-sdio.Raspberry Pi Foundation-Raspberry Pi Compute Module 4.txt.zst",
+            "/usr/lib/firmware/brcm/brcmfmac4356-pcie.Intel Corporation-CHERRYVIEW D1 PLATFORM.txt.zst",
+            "/usr/lib/firmware/brcm/brcmfmac4356-pcie.Xiaomi Inc-Mipad2.txt.zst",
```
[the remark this answers is not in the document]
Comment (Bownairo) [reply ic-os/defs.bzl]: I was trying to keep this as small as possible, but I can see if the full `brcm` folder is much bigger. This will be included in the later HostOS changes, anyway.

File: ic-os/guestos/context/Dockerfile [hunk 3]
```diff
@@ -96,6 +96,7 @@ RUN \
 RUN systemctl enable \
     chrony \
     nftables \
+    ssh \
```
Comment (nmattia) [inline ic-os/guestos/context/Dockerfile:99]: was this already enabled before?

File: ic-os/guestos/envs/dev-malicious/BUILD.bazel [hunk 4]
```diff
@@ -7,6 +7,7 @@ load("//ic-os/guestos:defs.bzl", "image_deps")
 #    or //ic-os/defs.bzl for the full list of targets.
 icos_build(
     name = "dev-malicious",
+    build_local_base_image = True,
```
Comment (nmattia) [inline ic-os/guestos/envs/dev-malicious/BUILD.bazel:10]: As far as I understand, this builds the base image from scratch instead of pulling it from the registry. Why is that necessary?

File: ic-os/guestos/context/Dockerfile (hunk 3 above)
Comment (Bownairo) [reply ic-os/guestos/context/Dockerfile:99]: In 24.04, it seems ssh is no longer enabled by default.

File: ic-os/guestos/envs/dev-malicious/BUILD.bazel (hunk 4 above)
Comment (Bownairo) [reply ic-os/guestos/envs/dev-malicious/BUILD.bazel:10]: While I've been testing on CI, I'm skipping over the registry at the loss of determinism, so that I'm not pushing a bunch of new images. When we're ready to roll, I'll need to break those changes to a separate PR to push the new image.

Comment (Bownairo) [reply ic-os/guestos/envs/dev-malicious/BUILD.bazel:10]: The later, this builds the base image directly in the pipeline rather than looking for it. The [`container-base-images.yml`](https://github.com/dfinity/ic/actions/workflows/container-base-images.yml) workflow pushes these regularly.

I've switched back to a draft until we're ready to push the image 👍.

**Comment by Bownairo** [comment]
> hi @Bownairo could you please merge master into your branch before running [Schedule Daily](https://github.com/dfinity/ic/actions/workflows/schedule-daily.yml) again?

Sorry, will do. Is this a one time thing or should I not be running this schedule? I'm mostly looking to trigger `bazel-test-bare-metal` and some of the performance tests.

File: ic-os/guestos/context/Dockerfile [hunk 5]
```diff
@@ -129,19 +126,8 @@ RUN mkdir -p /var/lib/ic/backup \
...
-## This is not reproducible. 
-## Reset the passwords to the locked password ("!!"). `usermod` honors SOURCE_DATE_EPOCH (set in the base dockerfile). This removes the password timestamp.
-## Why not use `faketime`? Doesn't work with `podman --timestamp=0`.
-RUN systemd-sysusers && \
-    usermod -p '!!' systemd-timesync && \
-    usermod -p '!!' systemd-coredump
-
 # Set /bin/sh to point to /bin/bash instead of the default /bin/dash
-RUN echo "set dash/sh false" | debconf-communicate && dpkg-reconfigure -fnoninteractive dash
+RUN ln -sf bash /usr/bin/sh
```
Comment (sasa-tomic) [inline ic-os/guestos/context/Dockerfile:130]:
```suggestion
RUN ln -sf /bin/bash /usr/bin/sh
```
